web/book/src/internals/modules.md

```diff
@@ -0,0 +1,324 @@
+declaration , which means that it can contain nested child modules.
+
+This means that modules form a
+[tree graph](<https://en.wikipedia.org/wiki/Tree_(graph_theory)>), which we call
+"the module structure".
+
+For the sake of this document, we will express the module structure with
+`module` keyword and a code block encased in curly braces:
+
+```
+module my_playlists {
+    let bangers = ... # a declaration
+
+    module soundtracks {
+        let movie_albums = ... # another declaration
+    }
+}
+```
+
+> The syntax `module name { ...decls... }` is not part of PRQL language, with
+> the objection that it is unnecessary as it only adds more ways of defining
+> modules. If a significant upside of this syntax is found, it may be added in
+> the future.
+
+## Name resolution
+
+Any declarations within a module can be referenced from the outside of the
+module:
+
+```prql no-eval
+# using module structure declared above
+
+let great_tracks = my_playlists.bangers
+
+let movie_scores = my_playlists.soundtracks.movie_albums
+```
+
+Identifiers are resolved relative to current module.
+
+```prql no-eval
+module my_playlists {
+    module soundtracks {
+        let movie_albums = (from albums | filter id == 3)
+    }
+
+    from soundtracks.movie_albums
+}
+from my_playlists.soundtracks.movie_albums
+```
+
+If an identifier cannot be resolved relative to the current module, it tries to
+resolve relative to the parent module. This is repeated, stepping up the module
+hierarchy until a match is found or root of the module structure is reached.
+
+```prql no-eval
+module my_playlists {
+    let decl_1 = ...
+
+    module soundtracks {
+        let decl_2 = ...
+    }
+
+    module upbeat_rock {
+        let decl_3 = ...
+
+        from decl_1 | join soundtracks.decl2 | join decl_3
+    }
+}
+```
+
+## Main var declaration
+
+If the final declaration in a module is a variable declaration that is named
+`main`, then the leading `let main = ` can be omitted and expressed only by the
+expression itself.
+
+```
+module my_playlists {
+    let bangers = (from tracks | take 10)
+
+    from playlists | take 10
+}
+
+let album_titles = my_playlists.main
+```
+
+When a module is referenced as a value, the `main` variable is used instead.
+This is especially useful when referring to a module which is to be compiled to
+RQ (and later SQL).
+
+```
+# last line from previous example could thus be shortened to:
+let album_titles = my_playlists
+```
+
+## File importing
+
+To include PRQL source code from other files, one can use the following syntax:
+
+```
+module my_playlists
+```
+
+Such declaration loads either `./my_playlists.prql` (a leaf module) or
+`./my_playlists/_my_playlists.prql` (a directory module) and uses its contents
+as the declarations in module `my_playlists`. If none or both of the files are
+present, a compilation error is raised.
+
+Only directory modules can contain module declarations. If a leaf module
```

Comment:
I don't get this, then (from https://github.com/PRQL/prql/pull/2129/files#r1161422818) — and [here](https://github.com/PRQL/prql/pull/2129/files#diff-4923b398dd490faaaf22d6d981064e8fda7d70906cfb1c3af2144666a03cc86aR289) there's an import (declaration?) in a leaf module? (I'm sure I'm misunderstanding...)

Reply:
OK, I _think_ I get the root of my confusion. What is a "module declaration"? 

Is it `module foo` or the module file? (I had thought this line was saying we could only put `module foo` in directory modules, since that's how it's referred to in the paragraph above)

I'll add a couple of suggestions to clear this up

Reply:
module declaration is `module foo`. And yes, that's what I meant: you can only put `module foo` into directory modules such as the one that is loaded from `sales/_sales.prql`.

If you'd put `module foo` into `sales/orders.prql`, you'd get this error:
```
module `sales.orders` is not a directory module, but it declares sub modules.

Hint: convert it to a directory module by moving file `sales/order.prql` to `sales/orders/_orders.prql`
```

Reply:
OK, I'm looking at the Example now — I should have started there, sorry for this noise — that's the clearest way of anchoring the discussion. 

I see the intention now, I'll comment there.